edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,275 @@
+            }
+        }
+
+        public void Remove(string id)
+        {
+            // In the case of an explicit removal, the device/module is
+            // no longer part of the nested hierarchy, so we should
+            // remove any children of the device as well.
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            var writableSnapshot = new Dictionary<string, ServiceIdentityTreeNode>(snapshot);
+
+            this.RemoveRecursive(writableSnapshot, id);
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling Remove(), so this should never fail
+                throw new InvalidOperationException("Multi-threaded insert/update detected on ServiceIdentityTree");
+            }
+        }
+
+        public bool Contains(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.ContainsKey(id);
+        }
+
+        public Option<ServiceIdentity> Get(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode)
+                    ? Option.Some(treeNode.Identity)
+                    : Option.None<ServiceIdentity>();
+        }
+
+        public IList<string> GetAllIds()
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.Select(kvp => kvp.Value.Identity.Id).ToList();
+        }
+
+        public bool TryGetAuthChain(string id, out string authChain)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            authChain = string.Empty;
+
+            // Auth-chain for root
+            if (this.rootDeviceId == id)
+            {
+                authChain = this.rootDeviceId;
+                return true;
+            }
+
+            // Auth-chain for a child somewhere in the tree
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            if (snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode))
+            {
+                // Auth chain starts with the originating device
+                authChain = treeNode.Identity.Id;
+
+                // Walk the parents up to the root
+                while (treeNode.Parent.HasValue)
+                {
+                    // Accumulate the auth chain per parent visited in the format:
+                    // "leaf1;edge1;edge2;...;rootEdge"
+                    ServiceIdentityTreeNode parent = treeNode.Parent.Expect(() => new InvalidOperationException());
+                    authChain += ";" + parent.Identity.Id;
```

Comment:
should we use local variable and assign authChain just before return true statement?